lib/experimental/add-registered-webfonts-to-theme-json.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Extend theme.json with programmatically registered webfonts.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Add missing fonts data to the global styles.
+ *
+ * @param array $data The global styles.
+ * @return array The global styles with missing fonts data.
+ */
+function gutenberg_add_registered_webfonts_to_theme_json( $data ) {
```

Comment:
If we go this route and only enqueue webfonts from theme.json & styles variations JSON files, then I'm not sure what the point of this function is anymore.
It made sense in the initial implementation where all webfonts were getting enqueued by default, but if we explicitly require theme & plugin authors to enqueue programmatically added webfonts, then is there a point to adding them in the theme.json array here? 🤔 

From what I understand, theme & plugin authors that register webfonts via PHP, will need to do an extra `wp_webfonts()-> enqueue_webfont()` call anyway, so perhaps this whole function should be completely removed, and theme/plugin authors should be left to handle enqueueing both in the front and the editor?

Reply:
Ah please ignore this comment... I forgot that we need to add the fonts to theme.json this way so that they appear in the font-family picker in the editor.
That is essentially its main purpose, not so much the webfonts registration/enqueing part.

Reply:
Exactly. However, what's the point of adding the fonts to the picker if they're not going to be enqueued based on being picked alone?

What if we add this functionality under a flag/filter? For example, the [Jetpack PR](https://github.com/Automattic/jetpack/pull/23810) will introspect blocks and global styles for fonts picked, but Gutenberg won't, at least for now until we migrate that feature from Jetpack to Gutenberg.

I'm thinking of a filter called `expose_registered_webfonts_to_block_editor`, or something like that.

Reply:
Something like that, @aristath:

```diff
diff --git a/lib/experimental/class-wp-theme-json-resolver-gutenberg.php b/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
index fb2416c7aa..9e9afe0dce 100644
--- a/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
+++ b/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
@@ -40,9 +40,7 @@ class WP_Theme_JSON_Resolver_Gutenberg extends WP_Theme_JSON_Resolver_6_0 {
 			$original_theme_json = new WP_Theme_JSON_Gutenberg( $theme_json_data );
 			gutenberg_register_webfonts_from_theme_json( $original_theme_json->get_settings() );
 
-			if ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
-				$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
-			}
+			$theme_json_data = static::maybe_add_registered_webfonts_to_theme_json( $theme_json_data );
 
 			static::$theme = new WP_Theme_JSON_Gutenberg( $theme_json_data );
 
@@ -106,4 +104,26 @@ class WP_Theme_JSON_Resolver_Gutenberg extends WP_Theme_JSON_Resolver_6_0 {
 		return $with_theme_supports;
 	}
 
+	/**
+	 * Add registered webfonts to theme.json if loading the block editor
+	 * and explicitly opted into showing them.
+	 *
+	 * This is for plugins that introspect blocks and global styles so
+	 * they could enqueue fonts picked in the block editor.
+	 *
+	 * @param array $theme_json_data The theme.json, represented as an array.
+	 *
+	 * @return array The theme.json with the fonts addded.
+	 */
+	private static function maybe_add_registered_webfonts_to_theme_json( $theme_json_data ) {
+		// If in the editor...
+		if ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
+			// And explicitly opted into showing the registered webfonts in the block editor...
+			if ( true === apply_filters( 'gutenberg_expose_registered_webfonts_to_block_editor', false ) ) {
+				return gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
+			}
+		}
+
+		return $theme_json_data;
+	}
 }
```

Reply:
`maybe_add_registered_webfonts_to_theme_json` is a terrible name, though :D It's not descriptive at all.

Something like `maybe_expose_registered_webfonts_in_block_editor` might be better. The same thing applies to `gutenberg_add_registered_webfonts_to_theme_json`... Yes, we're adding the webfonts to `theme.json`, but why? It's not obvious.

Reply:
Parsing blocks on-render to get the font-families and enqueue them can easily work on the frontend, that won't be an issue... after all, we do the same for the blocks' CSS styles and it works flawlessly.
If a user selects a font-family globally in their site-editor, then that font-family gets added as inline styles in `body{font-family: Something;}`. Ideally, the `body` would just get a class `has-font-family--whatever` just like all other elements, and we'd then be able to pick that up and enqueue the right font-family on the front more efficiently.
On the editor side, we can either just enqueue everything, OR lazy-load selected webfonts.

About lazy-loading:
Lazy-loading the selected webfont can be done by using a mechanism similar to what the [webfontloader](https://github.com/typekit/webfontloader) project was doing years ago. Of course we can't use that project since it's too old, unmaintained, and contains a bunch of stuff that we don't need... but the principle is sound. 
Basically we can write a small script that injects in the DOM the styles required for the webfonts. Writing that script is an easy task, however the only reason I haven't attempted to implement such a feature is because I wasn't sure if we want to hook in the font-family pickers themselves, or if we should create a more abstract implementation for hooking in onChange events for all controls, and then use that in the font-family pickers.
Hooking directly in the font-family pickers is the easy way and we can do it without any significant issues.
A more abstract solution would probably be more useful long-term looking at the big picture, but I have no idea where to start building that 😬 

As a sidenote, that was one of the main reasons why the initial implementation of the webfonts API only included webfonts registration and not enqueuing... The vision was to eventually detect the webfonts used, and automatically enqueue the ones that are selected and used.

> However, what's the point of adding the fonts to the picker if they're not going to be enqueued based on being picked alone? Remember that theme.json is the source of truth for enqueued fonts, not the block editor.

Maybe that should not be the case?
Thinking of user expectations and what they will want to do, it makes more sense to have the site-editor be the source of truth, and not `theme.json`... The `theme.json` file is a tool used by developers to expose things to the site-editor, define things and provide some initial styling, but the site-editor is the tool that end-users will eventually use.
With the end-user in mind, it makes a lot more sense to have the site-editor be the source of truth, and not `theme.json` which is nothing but a developer tool 🤷

Reply:
@aristath: https://github.com/WordPress/gutenberg/pull/39988#issuecomment-1099340232

Reply:
Please read: https://github.com/WordPress/gutenberg/pull/39988#issuecomment-1099511571

And https://github.com/WordPress/gutenberg/issues/40364!

Reply:
I think that would be best, yes. Maybe wrap it in `if ( ! function_exists( '_wp_enqueue_webfonts_listed_in_theme_json' ) ) {` too, just to be future-proof as well?